### PR TITLE
provides default dsn value for charset in factories

### PR DESCRIPTION
**Pull request by oqq**
If anyone trusts the default config (e.g. me), than this person could get bad experience with the mysql event store without this PR.

json_decode() would return null with an invalid json string, e.g. a not correct utf-8 encoded string.
Which would end up with `Assert\InvalidArgumentException: payload must be an array`.

Anyway.. if there is no reason why the default charset should not be utf-8, this PR should be merged before next stable release to prevent for bc breaks.

### CODE REVIEW

**Comment by prolic** [comment]
wrong for postgres - build fails

**Comment by prolic** [comment]
must be something like: `pg_connect("host=localhost options='--client_encoding=UTF8'");`

**Comment by oqq** [comment]
how about to provide the dsn from the entire factory instead of in the abstract one?
Maybe a getDsn method could help. Meanings?

**Comment by oqq** [comment]
Here is how doctrine get around this..
https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Driver/PDOPgSql/Driver.php#L56

**Comment by oqq** [comment]
I have added my proposal and have extended tests. Please review again @prolic .

File: src/Container/PostgresEventStoreFactory.php [hunk 1]
```diff
@@ -52,16 +52,35 @@ protected function eventStoreClassName(): string
...
+        }
+
+        return $dsn;
+    }
+
     public function defaultOptions(): iterable
     {
         return [
             'connection_options' => [
-                'driver' => 'pdo_pgsql',
```
Comment (prolic) [inline src/Container/PostgresEventStoreFactory.php]: Why is that removed?

File: tests/TestUtil.php [hunk 2]
```diff
@@ -45,6 +45,7 @@ public static function getConnection(): PDO
             $dsn .= 'dbname=' . $connectionParams['dbname'] . $separator;
             $dsn = rtrim($dsn);
             self::$connection = new PDO($dsn, $connectionParams['user'], $connectionParams['password']);
+            self::$connection->query("SET NAMES '".$connectionParams['charset']."'");
```
Comment (prolic) [inline tests/TestUtil.php]: Why not via connection options?

File: src/Container/AbstractEventStoreFactory.php [hunk 3]
```diff
@@ -94,15 +81,15 @@ public function __invoke(ContainerInterface $container): EventStore
...
-            $password = $config['connection_options']['password'];
-            $connection = new PDO($dsn, $user, $password);
+            $connection = new PDO(
+                $this->buildConnectionDsn($config['connection_options']),
+                $config['connection_options']['user'],
+                $config['connection_options']['password']
+            );
+
+            if (isset($config['connection_options']['charset'])) {
+                $connection->query("SET NAMES '".$config['connection_options']['charset']."'");
```
Comment (prolic) [inline src/Container/AbstractEventStoreFactory.php]: Why not use connection options?

File: src/Container/PostgresEventStoreFactory.php (hunk 1 above)
Comment (oqq) [inline src/Container/PostgresEventStoreFactory.php]: Is not needed anymore.

File: tests/TestUtil.php (hunk 2 above)
Comment (oqq) [inline tests/TestUtil.php]: I have not found a connection option which would work for postgre. Thats the way how doctrine is bypass this issue.

File: src/Container/AbstractEventStoreFactory.php (hunk 3 above)
Comment (oqq) [inline src/Container/AbstractEventStoreFactory.php]: The connection option for charset is used in mysql factory. This is a workaround for postgre.

**Comment by prolic** [comment]
Try options='--client_encoding=UTF8' for PostgreSQL

**Comment by oqq** [comment]
@prolic any reason why you have used a space as separator for postgre? It seems that postgre is fine with ';' as separator.

**Comment by prolic** [comment]
| @prolic any reason why you have used a space as separator for postgre? It seems that postgre is fine with ';' as separator.

As far as I know ; doesn't work for postgres.
